### ISSUE TITLE
How to load pkl pre-trained model from Theano ?

### ISSUE THREAD

**Issue by shimiaoli** [tiny-dnn/tiny-dnn#621]
I have a pre-trained model in pkl format which was trained using Theano.
Now I want to write a test program using tiny-dnn. How can I convert the Theano pre-trained model to some format can be loaded by tiny-dnn ? or how can I write the weights into tiny-dnn model ?

Thanks.

**Comment by edgarriba** [comment]
One solution can be using [hickle](https://github.com/telegraphic/hickle) to save parameters in python and then load them from C++ using this script
https://github.com/szagoruyko/functional-zoo/tree/master/cpp_parser

**Comment by shimiaoli** [comment]
Thank @edgarriba for the advice. I will try it.
Then I have another question here. How to access the weights of a model in tiny-dnn so that I can write the loaded parameters in ? 
For example: I write such a model in tiny-dnn:
network<sequential> net;
net << convolutional_layer<relu>(32, 32, 5, 3, 9)
<< average_pooling_layer<relu>(28, 28, 9, 2)
<< fully_connected_layer<tan_h>(14 * 14 * 9, 120)
<< fully_connected_layer<softmax>(120, 40);
How to change the parameters of the first filter in the first layer? I cannot find it in tiny-dnn's documentation. 
I know there is a function to read the weights: std::vector<vec_t*> weights = nn[i]->get_weights();
But I cannot find a corresponding function such as set_weights().

**Comment by edgarriba** [comment]
`nn[i]->weights() = your weights` should work
Note that we changed the API `get_weights -> weights`

**Comment by shimiaoli** [comment]
Hi, @edgarriba,

In tiny-dnn documentation 1.0.0a1
it is mentioned: 
get weight vector
std::vector<vec_t*> weights = nn[i]->get_weights();
Number of elements differs by layer types and settings. For example, in fully-connected layer with bias term, weights[0] represents weight matrix and weights[1] represents bias vector.

I am wondering how are the number of elements in cnn layer case.
If the nn[i] layer is a cnn layer, with 8 convolutional filters , each filter has size 5*5*6,
then does  weights[0] represent the 1st cnn filter weights, weight[1] represent the 2nd cnn filter weights, ... and weight[8] represent the bias vector ?

Many thanks.

**Comment by shimiaoli** [comment]
I have tested nn[i]->weights() = your weights,  but it does not work. 
Then I have to change the weights in the json file.

**Comment by edgarriba** [comment]
@shimiaoli could you poiint us to a spinet so that we can see how do you do it?

**Comment by shimiaoli** [comment]
@edgarriba 
Here is how I do it:
float W[128*128*1];
float b[8];
....//assign values to W and b
net[0]->weights()[0] = &W;
net[0]->weights()[1] = &b;

The code can be compiled and run. But no weights in the net model are changed after that.

**Comment by shimiaoli** [comment]
@edgarriba 
I have just found out the right way to do it from a piece of code in master/tiny_dnn/io/caffe/layer_factory_impl.h

 vec_t &w = *dst->weights()[0];
  vec_t &b = *dst->weights()[1];

  // fill weights
  for (int o = 0; o < out_channels; o++) {
    for (int i = 0; i < in_channels; i++) {
      if (!table.is_connected(o, i)) {
        dst_idx += window_size * window_size;
        continue;
      }
      for (int x = 0; x < window_size * window_size; x++) {
        // TODO
        // dst->weight()[dst_idx++] = weights.data(src_idx++);
        w[dst_idx++] = weights.data(src_idx++);
      }
    }
  }

thanks

**Comment by edgarriba** [comment]
rgiht, that's the proper way to do it. /cc @karandesai-96

**Comment by shimiaoli** [comment]
Just add a note here to anyone who will do similar work. Theano and tiny-dnn have different conventions in  CNN layer & dropout layer computation. One should be very careful while converting the models. In CNN layer, filter is flipped 180 degree in Theano before compute correlation. For dropout layer, output is scaled down by the drop-out factor in Theano while unchanged in tiny-dnn.

**Comment by karandesai-96** [comment]
Thanks for the point of caution @shimiaoli
